sbi/inference/trainers/base.py

```diff
@@ -1062,15 +904,655 @@ def __setstate__(self, state_dict: Dict):
+
+            # For each invalid row, select a random column to flip
+            cols_to_flip = torch.randint(
+                num_nodes, (invalid_indices[0].shape[0],), device=inputs.device
+            )
+
+            # Create full indices for flipping and apply the flip
+            indices_to_flip = invalid_indices + (cols_to_flip,)
+            condition_masks[indices_to_flip] = ~condition_masks[indices_to_flip]
+        return condition_masks
+
+    def _default_edge_masks_generator(self, inputs):
+        """The default edge mask generator employed
+        if none is specified by the user. It simply pass
+        None as a full-attention masks (equivalent to a
+        full ones) in order to save memory. `input` is specified
+        for compatibility altough ignored."""
+
+        return None
+
+    def build_conditional(
+        self,
+        condition_mask: Union[Tensor, list],
+        edge_mask: Optional[Tensor],
+        mvf_estimator: Optional[MaskedConditionalVectorFieldEstimator],
+        prior: Optional[Distribution],
+        sample_with: Literal['ode', 'sde'],
+        conditional_parameters: Optional[PosteriorParameters],
+        **kwargs,
+    ) -> NeuralPosterior:
+        r"""Method for building an arbitrary conditional.
+
+        This method serves as a base method for constructing a conditional based
+        on a given estimator and prior. The conditional can be sampled using one of
+        several inference methods specified by `sample_with`.
+
+        This method serves the NeuralPosterior object to provide the final conditional,
+        despite the name, the returned object will not (necessarely) represent a
+        posterior but the appropriate conditional defined by the fixed condition
+        mask passed.
+
+        Args:
+            condition_masks: A boolean mask indicating the role of each variable.
+            Expected shape: `(batch_size, num_variables)`.
+            - `True` (or `1`): The variable at this position is observed and its
+                features will be used for conditioning.
+            - `False` (or `0`): The variable at this position is latent and its
+                features are subject to inference.
+            estimator: The estimator that the conditional is based on.
+            prior: A probability distribution that expresses prior knowledge about the
+                parameters, e.g. which ranges are meaningful for them. Must be a PyTorch
+                distribution, see FAQ for details on how to use custom distributions.
+            sample_with: The inference method to use. Must be one of:
+                - "sde"
+                - "ode"
+            **kwargs: Additional method-specific parameters.
+
+        Returns:
+            NeuralPosterior object.
+        """
+
+        condition_mask = torch.as_tensor(condition_mask)
+        if edge_mask is not None:
+            edge_mask = torch.as_tensor(edge_mask)
+
+        prior = self._resolve_prior(prior)
```

Comment:
This will cause problems if prior is None, as you found out.

I talked with Jan about it. Currently e.g. NPE gets away with that because it uses `ImproperPrior` as a "filler" object.

This isnt that nice, and we previously ran into problems with nans comming from there.

So: We can create an **explicit** filler object, which is a distribution called e.g. `NoPrior` that implements `log_probs` to always return 0 and support to be real and raise an error on sample. I think this would be for now the best solution, but not sure. You can also add #1635 in a inline comment.